### PR TITLE
[13.0][ADD] ebill_paynet v2

### PR DESCRIPTION
**Pull request by TDu**
This is  to followup on #551.That PR included the `ebill_paynet_account_financial_discount` which needs to be open on its own #586.
And also the server_env PR #559 related to paynet is now included here.

### CODE REVIEW

File: ebill_paynet/models/ebill_payment_contract.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+    paynet_account_number = fields.Char(string="Paynet ID", size=20)
+    is_paynet_contract = fields.Boolean(
+        compute="_compute_is_paynet_contract", store=False
+    )
+    paynet_service_id = fields.Many2one(
+        comodel_name="paynet.service", string="Paynet Service", ondelete="restrict"
+    )
+    payment_type = fields.Selection(
+        selection=[("qr", "QR"), ("esr", "ESR")],
```
Comment (yvaucher) [inline ebill_paynet/models/ebill_payment_contract.py]:
```suggestion
        selection=[("qr", "QR"), ("isr", "ISR")],
```

File: ebill_paynet/models/paynet_invoice_message.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+        string="IC Ref", size=14, help="Document interchange reference"
+    )
+    # Set with invoice_id.number but also with returned data from server ?
+    ref = fields.Char("Reference N°", size=35)
+    ebill_account_number = fields.Char("Paynet Id", size=20)
+    payload = fields.Text("Payload sent")
+    response = fields.Text("Response recieved")
+    shipment_id = fields.Char(size=24, help="Shipment Id on Paynet service")
+    payment_type = fields.Selection(
+        selection=[("qr", "QR"), ("esr", "ESR"), ("esp", "ESP"), ("npy", "NPY")],
```
Comment (yvaucher) [inline ebill_paynet/models/paynet_invoice_message.py]:
```suggestion
        selection=[("qr", "QR"), ("isr", "ISR"), ("esp", "ESP"), ("npy", "NPY")],
```

File: ebill_paynet/messages/invoice-2003A.xml [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+            {%- if biller.street2 %}
+            <Line-35>{{ biller.street2|truncate(5, True, "") }}</Line-35>
+            {% endif %}
+          </STREET>
+          <City>{{ biller.city or ''}}</City>
+          <Zip>{{ biller.zip or '' }}</Zip>
+          <Country>{{ biller.country_id.code or 'CH' }}</Country>
+        </NAME-ADDRESS>
+        <BANK-INFO>
+          {%- if payment_type in ['esr', 'esp'] %}
```
Comment (yvaucher) [inline ebill_paynet/messages/invoice-2003A.xml]:
```suggestion
          {%- if payment_type in ['isr', 'esp'] %}
```

File: ebill_paynet/models/account_invoice.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        self.ensure_one()
+        contract = self.partner_id.get_active_contract(self.transmit_method_id)
+        if not contract:
+            return
+        # Generate PDf to be send
+        pdf_data = []
+        report_names = ["account.report_invoice"]
+        if contract.payment_type == "qr":
+            report_names.append("l10n_ch.qr_report_main")
+        elif contract.payment_type == "esr":
```
Comment (yvaucher) [inline ebill_paynet/models/account_invoice.py]:
```suggestion
        elif contract.payment_type == "isr":
```

File: ebill_paynet/messages/invoice-2013A.xml [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+            <Date>{{ format_date(invoice.date_invoice) }}</Date>
+          </REFERENCE-DATE>
+        </INVOICE-REFERENCE>
+      </REFERENCE>
+      <BILLER>
+        {#- Doc says vat number or business identitfiaction number #}
+        {%- if invoice.company_id.vat %}
+        <Tax-No>{{ invoice.company_id.vat }}</Tax-No>
+        {%- endif %}
+        {%- if invoice_esr %}
```
Comment (yvaucher) [inline ebill_paynet/messages/invoice-2013A.xml:39]:
```suggestion
        {%- if invoice_isr %}
```

File: ebill_paynet/messages/invoice-2013A.xml [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+            <Doc-Reference Type="ESR-NEU">{{ invoice_esr }}</Doc-Reference>
```
Comment (yvaucher) [inline ebill_paynet/messages/invoice-2013A.xml:40]:
```suggestion
            <Doc-Reference Type="ESR-NEU">{{ invoice_isr }}</Doc-Reference>
```

File: ebill_paynet/messages/invoice-2013A.xml (hunk 5 above)
Comment (yvaucher) [reply ebill_paynet/messages/invoice-2013A.xml:39]: Where does come this value? I don't see it anywhere.

**Review by yvaucher — CHANGES_REQUESTED** [review]
To be coherent with other module, please use the english term ISR and not ESR.

We changed from BVR to ISR to have less mixed terms, so always prefer ISR when possible. I suppose that the xml tags can only be ESR du to the ebill standard, but in Odoo all displayed values must be ISR.

File: ebill_paynet/models/paynet_invoice_message.py [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+        default="qr",
+        readonly=True,
+    )
```
Comment (yvaucher) [inline ebill_paynet/models/paynet_invoice_message.py:64]: a help text could help to give the full names of QR, ISR, ESP and NPY (I have no idea what the 2 latest are)

File: ebill_paynet/models/paynet_invoice_message.py [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+            ("reject", "Reject"),
+            ("error", "Error"),
+        ],
+        default="draft",
+    )
+    ic_ref = fields.Char(
+        string="IC Ref", size=14, help="Document interchange reference"
+    )
+    # Set with invoice_id.number but also with returned data from server ?
+    ref = fields.Char("Reference N°", size=35)
```
Comment (yvaucher) [inline ebill_paynet/models/paynet_invoice_message.py]:
```suggestion
    ref = fields.Char("Reference No.", size=35)
```
`N°` is for french localization 

For english you will want to use `No.`

https://en.wikipedia.org/wiki/Numero_sign

Comment (TDu) [reply ebill_paynet/models/paynet_invoice_message.py]: Fixed

File: ebill_paynet/messages/invoice-2013A.xml (hunk 6 above)
Comment (TDu) [reply ebill_paynet/messages/invoice-2013A.xml:40]: This file is related to Paynet B2C service and is not implemented yet but kept for further reference

File: ebill_paynet/messages/invoice-2013A.xml (hunk 5 above)
Comment (TDu) [reply ebill_paynet/messages/invoice-2013A.xml:39]: Same than below

File: ebill_paynet/models/account_invoice.py (hunk 4 above)
Comment (TDu) [reply ebill_paynet/models/account_invoice.py]: Changed

File: ebill_paynet/messages/invoice-2003A.xml (hunk 3 above)
Comment (TDu) [reply ebill_paynet/messages/invoice-2003A.xml]: Changed

File: ebill_paynet/models/paynet_invoice_message.py (hunk 2 above)
Comment (TDu) [reply ebill_paynet/models/paynet_invoice_message.py]: Changed

File: ebill_paynet/models/ebill_payment_contract.py (hunk 1 above)
Comment (TDu) [reply ebill_paynet/models/ebill_payment_contract.py]: Changed

**Comment by TDu** [comment]
@yvaucher I did the changes regarding the naming of esr/isr.
But the build is red and it seems to be related to another module, would you have any idea why ? Changes in the core ?

**Comment by TDu** [comment]
@yvaucher Regarding the broken tests, I opened this https://github.com/OCA/l10n-switzerland/pull/603

File: ebill_paynet/__manifest__.py [hunk 9]
```diff
@@ -17,6 +17,7 @@
         "l10n_ch_qriban",
         "queue_job",
         "sale",
+        "delivery",
```
Comment (simahawk) [inline ebill_paynet/__manifest__.py:20]: wouldn't be better to add a glue module?

Comment (TDu) [reply ebill_paynet/__manifest__.py:20]: Yes I think it would be...

**Comment by yvaucher** [comment]
some rebase to do

File: ebill_paynet/__manifest__.py (hunk 9 above)
Comment (TDu) [reply ebill_paynet/__manifest__.py:20]: I have added a comment in the roadmap about it.

**Comment by TDu** [comment]
It has been rebased.

**Comment by TDu** [comment]
@yvaucher Rebased done, and Travis is :green_heart:

File: ebill_paynet/components/api.py [hunk 10]
```diff
@@ -0,0 +1,58 @@
+# Copyright 2019 Camptocamp SA
```
Comment (simahawk) [inline ebill_paynet/components/api.py:1]: why `components` if there's no component here?

**Review by simahawk — APPROVED** [review]
LG overall. Could you add to ROADMAP "consider refactoring on top of EDI framework for v14 or v15"?

File: ebill_paynet/components/api.py (hunk 10 above)
Comment (TDu) [reply ebill_paynet/components/api.py:1]: No components in the sense of https://github.com/OCA/connector/tree/13.0/component, true.

**Comment by TDu** [comment]
> LG overall. Could you add to ROADMAP "consider refactoring on top of EDI framework for v14 or v15"?

Done

**Review by yvaucher — APPROVED** [review]

**Pull request merged**